src/docs/pantsd.md

```diff
@@ -0,0 +1,77 @@
+Pants Daemon
+============
+
+ProcessManager
+--------------
+
+`ProcessManager` is a class designed to keep track of processes. Besides changing their state (paused, terminating...), it allows a process to fork in different ways. Classes that extend `ProcessManager` can (and often do) inject code to be run before and after forking by overriding the functions `pre_fork`, `post_fork_child`, `post_fork_parent`. Some examples of usage of these functions are `DaemonPantsRunner` invoking a `LocalPantsRunner` in its `post_fork_child`, to fulfill a request to `PailgunServer`.
+
+The `PantsDaemon` Class
+-----------------------
+
+The [`PantsDaemon`]() class is responsible for managing the lifetime of the pantsd process and the services associated with it. A `PantsDaemon` is created with the inner class `Factory`, and it has two modes of initialization:
+
+- Stub initialization, which parses options, launches watchman, and not much else. This mode is used to determine whether it needs to fully initialize the daemon or not.
+- Full initialization, which spawns the daemon process, initializes all the services, the legacy engine and the native code.
+
+Initialization is encapsulated in the `PantsDaemon.Factory.create()` method.
+`PantsDaemon` is a `ProcessManager`, which means one can know if it's alive, or if it needs to restart.
+
+`PantsDaemon`s can be `launch()`ed, which will terminate the process it was running, and call `daemon_spawn()` to fork a new process. This new process will run the code in `PantsDaemon.post_fork_child()`, which in short means it will run `os.spawnve` to execute the `pants_daemon.py:launch()` function, which will call `PantsDaemon.run_sync()`. `run_sync()` does a lot of things, but the vital things are calling `_setup_services()` to spin up services, and `_run_services()` to start an infinte loop polling them. 
+
+Pailgun
+-------
+
+The [Nailgun Protocol]() is a protocol designed to allow clients to make command-line requests to a [Nailgun Server](). It supports an interface similar to [`ExecuteProcessRequest`](), except for it having no concern about making operations hermetic. Pailgun is an extension of the Nailgun Protocol, which is clients use to ask the Pants Daemon to spawn pants invocations.
```

Comment:
It's an important distinction that the nailgun protocol supports streaming access to stdin/stdout.